# Import necessary libraries

In [20]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder


In [27]:
# !pip install xgboost

Lab | Final regression model in "Health Care for All" Case

At this point, we have created a model to predict who will make a donation and who won't. But, what about the ammount of money that each person will give? In this lab, subset those that made a donation and use that subset to create a model to predict how much money will they give.

Evaluate the result of your model and estimate how much better the result are for the bussiness in comparison with the naive scenario we discuss on Monday.

# Get data

In [2]:
categorical = pd.read_csv('categorical.csv')
numerical = pd.read_csv('numerical.csv')
target = pd.read_csv('target.csv')

In [4]:
data = pd.concat([categorical,numerical,target],axis=1)
data.head()

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2 DOMAIN_A  \
0    IL       36        H      F         3      L      E        C        T   
1    CA       14        H      M         3      L      G        A        S   
2    NC       43        U      M         3      L      E        C        R   
3    CA       44        U      F         3      L      E        C        R   
4    FL       16        H      F         3      L      F        A        S   

   DOMAIN_B  ...  MAXRAMNT  LASTGIFT  TIMELAG    AVGGIFT  CONTROLN  HPHONE_D  \
0         2  ...      12.0      10.0        4   7.741935     95515         0   
1         1  ...      25.0      25.0       18  15.666667    148535         0   
2         2  ...      16.0       5.0       12   7.481481     15078         1   
3         2  ...      11.0      10.0        9   6.812500    172556         1   
4         2  ...      15.0      15.0       14   6.864865      7112         1   

   RFA_2F  CLUSTER2  TARGET_B  TARGET_D  
0       4        39         0       0.0  
1       2         1         0       0.0  
2       4        60         0       0.0  
3       4        41         0       0.0  
4       2        26         0       0.0  

[5 rows x 339 columns]

In [5]:
target_d = data[data['TARGET_B']==1]

In [6]:
target_d.shape

(4843, 339)

In [7]:
target_d.head()

STATE  CLUSTER HOMEOWNR GENDER  DATASRCE RFA_2R RFA_2A GEOCODE2 DOMAIN_A  \
20  other       12        H      F         3      L      D        A        S   
30     TX       35        H      M         3      L      D        A        T   
45  other       24        H      F         3      L      D        C        C   
78     CA       13        H      F         2      L      F        A        S   
93     GA       18        H      M         3      L      E        A        S   

    DOMAIN_B  ...  MAXRAMNT  LASTGIFT  TIMELAG    AVGGIFT  CONTROLN  HPHONE_D  \
20         1  ...       7.0       5.0       12   4.066667     82943         1   
30         1  ...      10.0       7.0        9   6.181818    190313         1   
45         1  ...       6.0       5.0        3   4.857143     76585         1   
78         1  ...      17.0      10.0       21  11.000000    156378         0   
93         2  ...      12.0      12.0        6   9.400000     25641         1   

    RFA_2F  CLUSTER2  TARGET_B  TARGET_D  
20       3         3         1       4.0  
30       3        14         1       7.0  
45       3        11         1       5.0  
78       2         2         1      13.0  
93       3        22         1      10.0  

[5 rows x 339 columns]

# X-y Split

In [9]:
X = target_d.drop(['TARGET_B','TARGET_D'],axis=1)
y = target_d['TARGET_D']

# Train-test split

In [10]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [11]:
print('X_train shape: ', X_train.shape)
print('X_test shape: ', X_test.shape)
print('y_train shape: ', y_train.shape)
print('y_test shape: ', y_test.shape)

X_train shape:  (3874, 337)
X_test shape:  (969, 337)
y_train shape:  (3874,)
y_test shape:  (969,)


In [12]:
X_train_num = X_train.select_dtypes(include = np.number)
X_test_num  = X_test.select_dtypes(include = np.number)
X_train_cat = X_train.select_dtypes(include = object)
X_test_cat  = X_test.select_dtypes(include = object)

In [13]:
print('X_train_num shape: ', X_train_num.shape)
print('X_test_num shape: ', X_test_num.shape)
print('X_train_cat shape: ', X_train_cat.shape)
print('X_test_cat shape: ', X_test_cat.shape)

X_train_num shape:  (3874, 330)
X_test_num shape:  (969, 330)
X_train_cat shape:  (3874, 7)
X_test_cat shape:  (969, 7)


In [16]:
transformer = MinMaxScaler().fit(X_train_num)
X_train_num_norm = transformer.transform(X_train_num)
X_train_num_scale = pd.DataFrame(X_train_num_norm, index = X_train_num.index, columns=X_train_num.columns)
X_train_num_scale.head()
X_train_num_scale.reset_index(drop = True, inplace = True)

In [18]:
X_test_num_norm = transformer.transform(X_test_num)
X_test_num_scale = pd.DataFrame(X_test_num_norm, index = X_test_num.index, columns=X_test_num.columns)
X_test_num_scale.head()
X_test_num_scale.reset_index(drop = True, inplace=True)
# X_test_num_scale

In [21]:
encoder = OneHotEncoder(drop='first').fit(X_train_cat)

cols = encoder.get_feature_names(input_features=X_train_cat.columns)

X_train_cat_encode = pd.DataFrame(encoder.transform(X_train_cat).toarray(),columns=cols)

X_train_cat_encode.reset_index(drop = True, inplace = True)
X_train_cat_encode

/Users/ozguryildirim/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


STATE_FL  STATE_GA  STATE_IL  STATE_IN  STATE_MI  STATE_MO  STATE_NC  \
0          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
4          0.0       0.0       0.0       0.0       0.0       0.0       0.0   
...        ...       ...       ...       ...       ...       ...       ...   
3869       0.0       0.0       0.0       0.0       0.0       0.0       1.0   
3870       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3871       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3872       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3873       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

      STATE_TX  STATE_WA  STATE_WI  ...  RFA_2A_E  RFA_2A_F  RFA_2A_G  \
0          0.0       0.0       0.0  ...       1.0       0.0       0.0   
1          0.0       0.0       0.0  ...       0.0       1.0       0.0   
2          0.0       0.0       0.0  ...       0.0       0.0       1.0   
3          0.0       1.0       0.0  ...       0.0       1.0       0.0   
4          0.0       1.0       0.0  ...       0.0       1.0       0.0   
...        ...       ...       ...  ...       ...       ...       ...   
3869       0.0       0.0       0.0  ...       0.0       0.0       1.0   
3870       0.0       0.0       0.0  ...       0.0       1.0       0.0   
3871       0.0       0.0       0.0  ...       0.0       1.0       0.0   
3872       0.0       0.0       0.0  ...       0.0       1.0       0.0   
3873       0.0       0.0       0.0  ...       0.0       1.0       0.0   

      GEOCODE2_B  GEOCODE2_C  GEOCODE2_D  DOMAIN_A_R  DOMAIN_A_S  DOMAIN_A_T  \
0            0.0         0.0         0.0         0.0         0.0         1.0   
1            1.0         0.0         0.0         0.0         1.0         0.0   
2            0.0         0.0         1.0         0.0         0.0         1.0   
3            0.0         1.0         0.0         0.0         0.0         0.0   
4            0.0         0.0         0.0         0.0         0.0         1.0   
...          ...         ...         ...         ...         ...         ...   
3869         1.0         0.0         0.0         0.0         0.0         0.0   
3870         0.0         1.0         0.0         0.0         0.0         0.0   
3871         1.0         0.0         0.0         0.0         0.0         1.0   
3872         1.0         0.0         0.0         0.0         1.0         0.0   
3873         0.0         0.0         1.0         1.0         0.0         0.0   

      DOMAIN_A_U  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  
...          ...  
3869         0.0  
3870         0.0  
3871         0.0  
3872         0.0  
3873         0.0  

[3874 rows x 24 columns]

In [22]:
cols = encoder.get_feature_names(input_features=X_test_cat.columns)

X_test_cat_encode = pd.DataFrame(encoder.transform(X_test_cat).toarray(),columns=cols)

X_test_cat_encode.reset_index(drop = True, inplace = True)
X_test_cat_encode

/Users/ozguryildirim/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


STATE_FL  STATE_GA  STATE_IL  STATE_IN  STATE_MI  STATE_MO  STATE_NC  \
0         0.0       0.0       0.0       0.0       0.0       0.0       0.0   
1         0.0       0.0       0.0       0.0       0.0       0.0       0.0   
2         0.0       0.0       0.0       0.0       0.0       0.0       0.0   
3         1.0       0.0       0.0       0.0       0.0       0.0       0.0   
4         0.0       0.0       0.0       1.0       0.0       0.0       0.0   
..        ...       ...       ...       ...       ...       ...       ...   
964       0.0       0.0       0.0       1.0       0.0       0.0       0.0   
965       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
966       0.0       0.0       0.0       0.0       1.0       0.0       0.0   
967       0.0       0.0       0.0       0.0       0.0       0.0       0.0   
968       0.0       0.0       0.0       0.0       0.0       0.0       0.0   

     STATE_TX  STATE_WA  STATE_WI  ...  RFA_2A_E  RFA_2A_F  RFA_2A_G  \
0         0.0       0.0       0.0  ...       0.0       1.0       0.0   
1         0.0       0.0       0.0  ...       0.0       1.0       0.0   
2         0.0       0.0       0.0  ...       0.0       1.0       0.0   
3         0.0       0.0       0.0  ...       0.0       0.0       0.0   
4         0.0       0.0       0.0  ...       0.0       0.0       0.0   
..        ...       ...       ...  ...       ...       ...       ...   
964       0.0       0.0       0.0  ...       0.0       1.0       0.0   
965       0.0       0.0       0.0  ...       0.0       1.0       0.0   
966       0.0       0.0       0.0  ...       0.0       0.0       1.0   
967       0.0       0.0       0.0  ...       0.0       1.0       0.0   
968       0.0       0.0       0.0  ...       0.0       1.0       0.0   

     GEOCODE2_B  GEOCODE2_C  GEOCODE2_D  DOMAIN_A_R  DOMAIN_A_S  DOMAIN_A_T  \
0           1.0         0.0         0.0         0.0         0.0         0.0   
1           0.0         1.0         0.0         1.0         0.0         0.0   
2           0.0         0.0         1.0         1.0         0.0         0.0   
3           1.0         0.0         0.0         0.0         0.0         0.0   
4           0.0         0.0         1.0         1.0         0.0         0.0   
..          ...         ...         ...         ...         ...         ...   
964         0.0         0.0         1.0         0.0         0.0         1.0   
965         0.0         1.0         0.0         0.0         0.0         1.0   
966         0.0         0.0         1.0         1.0         0.0         0.0   
967         0.0         0.0         0.0         0.0         1.0         0.0   
968         0.0         0.0         0.0         0.0         1.0         0.0   

     DOMAIN_A_U  
0           1.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  
..          ...  
964         0.0  
965         0.0  
966         0.0  
967         0.0  
968         0.0  

[969 rows x 24 columns]

In [23]:
X_train = pd.concat([X_train_num_scale,X_train_cat_encode], axis=1)
X_train.index
y_train.reset_index(drop = True, inplace = True)

In [24]:
X_test = pd.concat([X_test_num_scale,X_test_cat_encode], axis=1)
X_test

CLUSTER  DATASRCE  DOMAIN_B  ODATEW_YR  ODATEW_MM    DOB_YR    DOB_MM  \
0    0.000000       0.5  0.000000   0.909091        0.0  0.351648  0.000000   
1    0.865385       0.5  0.333333   0.000000        0.0  0.000000  0.090909   
2    0.846154       1.0  0.333333   0.818182        0.0  0.527473  0.000000   
3    0.538462       1.0  0.333333   0.000000        0.0  0.164835  0.090909   
4    0.846154       1.0  0.333333   0.818182        0.0  0.461538  0.000000   
..        ...       ...       ...        ...        ...       ...       ...   
964  0.673077       1.0  0.333333   0.636364        0.0  0.318681  1.000000   
965  0.653846       1.0  0.000000   0.545455        0.0  0.000000  0.090909   
966  1.000000       1.0  0.666667   0.545455        0.0  0.626374  0.000000   
967  0.250000       1.0  0.000000   0.363636        0.0  0.263736  0.000000   
968  0.346154       1.0  0.333333   0.818182        0.0  0.000000  0.090909   

     MINRDATE_YR  MINRDATE_MM  MAXRDATE_YR  ...  RFA_2A_E  RFA_2A_F  RFA_2A_G  \
0       0.909091     0.000000     0.909091  ...       0.0       1.0       0.0   
1       0.090909     0.454545     0.818182  ...       0.0       1.0       0.0   
2       0.727273     1.000000     0.818182  ...       0.0       1.0       0.0   
3       0.181818     1.000000     0.545455  ...       0.0       0.0       0.0   
4       0.818182     0.909091     0.818182  ...       0.0       0.0       0.0   
..           ...          ...          ...  ...       ...       ...       ...   
964     0.636364     0.090909     0.727273  ...       0.0       1.0       0.0   
965     0.545455     0.818182     0.818182  ...       0.0       1.0       0.0   
966     0.545455     0.363636     0.909091  ...       0.0       0.0       1.0   
967     0.363636     0.090909     0.818182  ...       0.0       1.0       0.0   
968     0.818182     0.090909     0.818182  ...       0.0       1.0       0.0   

     GEOCODE2_B  GEOCODE2_C  GEOCODE2_D  DOMAIN_A_R  DOMAIN_A_S  DOMAIN_A_T  \
0           1.0         0.0         0.0         0.0         0.0         0.0   
1           0.0         1.0         0.0         1.0         0.0         0.0   
2           0.0         0.0         1.0         1.0         0.0         0.0   
3           1.0         0.0         0.0         0.0         0.0         0.0   
4           0.0         0.0         1.0         1.0         0.0         0.0   
..          ...         ...         ...         ...         ...         ...   
964         0.0         0.0         1.0         0.0         0.0         1.0   
965         0.0         1.0         0.0         0.0         0.0         1.0   
966         0.0         0.0         1.0         1.0         0.0         0.0   
967         0.0         0.0         0.0         0.0         1.0         0.0   
968         0.0         0.0         0.0         0.0         1.0         0.0   

     DOMAIN_A_U  
0           1.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  
..          ...  
964         0.0  
965         0.0  
966         0.0  
967         0.0  
968         0.0  

[969 rows x 354 columns]

In [29]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from xgboost import XGBRegressor
from sklearn.ensemble import RandomForestRegressor

DT = DecisionTreeRegressor()
LR = LinearRegression()
xgb = XGBRegressor()
RF = RandomForestRegressor()

model_pipeline = [DT, LR, xgb, RF]
model_names = ['Regression Tree', 'Linear Regression','XGBoost','Random Forest']
scores = {}
i=0
for model in model_pipeline:
    mean_score = np.mean(cross_val_score(model, X_train, y_train, cv=10))
    scores[model_names[i]] = mean_score
    i = i+1
print(scores)


{'Regression Tree': -0.15300084887436716, 'Linear Regression': 0.4117895634825759, 'XGBoost': 0.43296158914105465, 'Random Forest': 0.5135369956701247}


In [30]:
val_scores = {}
i=0
for model in model_pipeline:
    model.fit(X_train, y_train)
    val_scores[model_names[i]] = model.score(X_test,y_test)
    i = i+1
print(val_scores)

{'Regression Tree': -0.10406409781674641, 'Linear Regression': 0.2895529501344797, 'XGBoost': 0.28323746181123277, 'Random Forest': 0.387041246847988}


In [31]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from math import sqrt

LR =RandomForestRegressor().fit(X_train, y_train)
pred = LR.predict(X_test)


In [32]:
r2 = r2_score(y_test,pred)
mse = mean_squared_error(y_test,pred)
mae = mean_absolute_error(y_test,pred)
rmse = sqrt(mse)

In [33]:
print('R2 Score:', r2)
print('Mean Squared Error:', mse)
print('Mean aboslute Error:', mae)
print('Root Mean Squared Error:',rmse)

R2 Score: 0.40136900370274187
Mean Squared Error: 114.07792042379772
Mean aboslute Error: 4.783957275541796
Root Mean Squared Error: 10.680726586885264


In [34]:
# how much will the donators give?
round(np.mean(pred),2)

16.36

As above shows we can expect that the average donation amount is 16.36 dollars. 

When you have a look at the calculations I made in the previous lab and update them with predicted donation amount, we get the following:

- Donation amount gained: 183820.96
- Donation amount lost: 114814.48
- Amount spend on marketing: 12449.44
- Amount lost on marketing: 4808.96

Looking at these numbers, we can see that we gain 183.820 dollars with the current model and we soend 12.449. Which means we have a donation gain of 171.371 dollars.
If we wouldn't have made a prediction and instead have send marketing efforts to all of all the 36.228 people (I am using the numbers in the test set), we would have spend 24.635 on marketing efforts and gained 298.635 dollars,leaving a total of 274.000 donations. This actually shows us that it will give us more money to send the marketing efforts to all the possible donators instead of following the model. However, I have not been able to improve the model just yet, so when I have time this weekend I will go back to it and improve the model, so that we lose less donators but also send out less marketing packages. 